daemon: Use TestMain, SetUpSuite, and SetUpTest

jrajahalme opened the pull request:
Separate initialization code that should only be executed once
into TestMain(). Set up the kv-store in SetUpSuite(), as the
kv-store setup does not need not be executed again for each test.

Keep most of the initialization and clean-up code in
SetUpTest()/TearDownTest() so that any changes made to the daemon, for
example, do not leak from a test to another.

This should fix panics like:
```
level=panic msg="Close() called without calling InitIdentityAllocator() first" subsys=identity-cache
----------------------------------------------------------------------
PANIC: daemon_test.go:173: DaemonEtcdSuite.TearDownTest
... Panic: &{%!s(*logrus.Logger=&{0xc000010020 map[2:[0xc000363430] 3:[0xc000363430]] 0xc0000dc730 false 4 {{0 0} false} {{} 0xc000579c00 2 <nil>} 0x12d6f70}) map[subsys:identity-cache] 2019-07-17 16:52:52.111645944 +0000 UTC m=+0.085771897 panic %!s(*runtime.Frame=<nil>) Close() called without calling InitIdentityAllocator() first <nil> <nil> } (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:239
  in Entry.log
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:268
  in Entry.Log
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:306
  in Entry.Panic
/home/travis/gopath/src/github.com/cilium/cilium/pkg/identity/cache/allocator.go:180
  in Close
daemon_test.go:152
  in DaemonSuite.TearDownTest
daemon_test.go:174
  in DaemonEtcdSuite.TearDownTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: <autogenerated>:1: DaemonEtcdSuite.TestEndpointAddInvalidLabel
... Panic: Fixture has panicked (see related PANIC)
level=info msg="Waiting for consul to elect a leader" subsys=kvstore
----------------------------------------------------------------------
PANIC: daemon_test.go:183: DaemonConsulSuite.SetUpTest
... Panic: duplicate metrics collector registration attempted (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/prometheus/client_golang/prometheus/registry.go:391
  in Registry.MustRegister
/home/travis/gopath/src/github.com/cilium/cilium/pkg/metrics/metrics.go:1040
  in MustRegister
/home/travis/gopath/src/github.com/cilium/cilium/pkg/option/config.go:1641
  in DaemonConfig.Populate
daemon_test.go:81
  in DaemonSuite.SetUpTest
daemon_test.go:185
  in DaemonConsulSuite.SetUpTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: daemon_test.go:188: DaemonConsulSuite.TearDownTest
... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:82
  in panicmem
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/signal_unix.go:390
  in sigpanic
daemon.go:996
  in Daemon.Close
daemon_test.go:154
  in DaemonSuite.TearDownTest
daemon_test.go:189
  in DaemonConsulSuite.TearDownTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: <autogenerated>:1: DaemonConsulSuite.TestEndpointAddInvalidLabel
... Panic: Fixture has panicked (see related PANIC)
```
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8601)
<!-- Reviewable:end -->

raybejjani commented on daemon/daemon_test.go:
```diff
@@ -77,19 +77,17 @@ type DaemonSuite struct {
 	tempRunDir, err := ioutil.TempDir("", "cilium-test-run")
-	c.Assert(err, IsNil)
+	if err != nil {
+		panic("TempDir() failed.")
```
You should be able to pass a third (or more) parameter to `c.Assert(err, IsNil)` that is treated as a loggable message. That would also include the error itself, unless you wanted to not print that?

jrajahalme replied:
'TestMain()' does not have 'c', so no 'Assert' either. Figured that if these fail we could just panic instead.

raybejjani commented on daemon/daemon_test.go:
```diff
@@ -77,19 +77,17 @@ type DaemonSuite struct {
 }
 
-func (ds *DaemonSuite) SetUpTest(c *C) {
+func TestMain(m *testing.M) {
```
Is `TestMain` something called by GoCheck? (probably?)

ianvernon replied:
Yeah, I find it a bit weird that we are mixing up `go-check` and the default testing library in Golang here.

@raybejjani re: `TestMain` - https://golang.org/pkg/testing/#hdr-Main

raybejjani replied:
ah, TIL... this answers my other comment too, I think!

jrajahalme replied:
@ianvernon `go-check` is explicitly built as an extension to the default `testing` library. Hence it is OK in my book to use `testing` functionality when `go-check` does not provide similar functionality. In this case it is about simply running global initialization code once before any testing is done. I like `TestMain()` better than, e.g., using `sync.Once` in, say, `SetUpSuite()`.

raybejjani commented on daemon/daemon_test.go:
```diff
@@ -100,14 +98,23 @@ func (ds *DaemonSuite) SetUpTest(c *C) {
 	// state left on disk.
 	option.Config.EnableHostIPRestore = false
 
+	os.Exit(m.Run())
```
Won't this cause the tests to exit? Or is it in a subprocess and I missed that?

raybejjani replied:
answered in the discussion around TestMain above

jrajahalme committed:
daemon: Use TestMain, SetUpSuite, and SetUpTest

Separate initialization code that should only be executed once
into TestMain(). Set up the kv-store in SetUpSuite(), as the
kv-store setup does not need not be executed again for each test.

Keep most of the initialization and clean-up code in
SetUpTest()/TearDownTest() so that any changes made to the daemon, for
example, do not leak from a test to another.

This should fix panics like:

level=panic msg="Close() called without calling InitIdentityAllocator() first" subsys=identity-cache
----------------------------------------------------------------------
PANIC: daemon_test.go:173: DaemonEtcdSuite.TearDownTest
... Panic: &{%!s(*logrus.Logger=&{0xc000010020 map[2:[0xc000363430] 3:[0xc000363430]] 0xc0000dc730 false 4 {{0 0} false} {{} 0xc000579c00 2 <nil>} 0x12d6f70}) map[subsys:identity-cache] 2019-07-17 16:52:52.111645944 +0000 UTC m=+0.085771897 panic %!s(*runtime.Frame=<nil>) Close() called without calling InitIdentityAllocator() first <nil> <nil> } (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:239
  in Entry.log
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:268
  in Entry.Log
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/entry.go:306
  in Entry.Panic
/home/travis/gopath/src/github.com/cilium/cilium/pkg/identity/cache/allocator.go:180
  in Close
daemon_test.go:152
  in DaemonSuite.TearDownTest
daemon_test.go:174
  in DaemonEtcdSuite.TearDownTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: <autogenerated>:1: DaemonEtcdSuite.TestEndpointAddInvalidLabel
... Panic: Fixture has panicked (see related PANIC)
level=info msg="Waiting for consul to elect a leader" subsys=kvstore
----------------------------------------------------------------------
PANIC: daemon_test.go:183: DaemonConsulSuite.SetUpTest
... Panic: duplicate metrics collector registration attempted (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/prometheus/client_golang/prometheus/registry.go:391
  in Registry.MustRegister
/home/travis/gopath/src/github.com/cilium/cilium/pkg/metrics/metrics.go:1040
  in MustRegister
/home/travis/gopath/src/github.com/cilium/cilium/pkg/option/config.go:1641
  in DaemonConfig.Populate
daemon_test.go:81
  in DaemonSuite.SetUpTest
daemon_test.go:185
  in DaemonConsulSuite.SetUpTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: daemon_test.go:188: DaemonConsulSuite.TearDownTest
... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x126CAF4)
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:522
  in gopanic
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/panic.go:82
  in panicmem
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/signal_unix.go:390
  in sigpanic
daemon.go:996
  in Daemon.Close
daemon_test.go:154
  in DaemonSuite.TearDownTest
daemon_test.go:189
  in DaemonConsulSuite.TearDownTest
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/reflect/value.go:308
  in Value.Call
/home/travis/.gimme/versions/go1.12.7.linux.amd64/src/runtime/asm_amd64.s:1337
  in goexit
----------------------------------------------------------------------
PANIC: <autogenerated>:1: DaemonConsulSuite.TestEndpointAddInvalidLabel
... Panic: Fixture has panicked (see related PANIC)

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
